cmake/SetupDoxygen.cmake

```diff
@@ -8,31 +8,99 @@ if (DOXYGEN_FOUND)
   set(SPECTRE_DOX_GENERATE_HTML "YES")
   set(SPECTRE_DOX_GENERATE_XML "NO")
   configure_file(
-      docs/Doxyfile.in
-      ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml @ONLY IMMEDIATE
+    docs/Doxyfile.in
+    ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml @ONLY IMMEDIATE
   )
   # Configure file that contains doxygen groups
   configure_file(docs/GroupDefs.hpp ${SPECTRE_DOXYGEN_GROUPS})
 
-  add_custom_target(
-      doc
-      COMMAND ${DOXYGEN_EXECUTABLE} ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml 2>&1
-      DEPENDS
-      ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml
-      ${SPECTRE_DOXYGEN_GROUPS}
+  # Construct the command that calls Doxygen
+  set(
+    GENERATE_DOCS_COMMAND
+    "${DOXYGEN_EXECUTABLE} ${PROJECT_BINARY_DIR}/docs/DoxyfileHtml"
+  )
+  # Construct the command that calls doxygen, but only outputs warnings with a
+  # few lines of context to stderr. Fails with exit code 1 if warnings are
+  # found. Remains silent and succeeds with exit code 0 if everything is fine.
+  # We write this into a shell script to make it easier to append the
+  # postprocessing command.
+  set(
+    GENERATE_AND_CHECK_DOCS_SCRIPT "\
+#!/bin/sh\n\
+! ${GENERATE_DOCS_COMMAND} 2>&1 | grep -A 6 'warning' >&2\n"
+  )
+
+  # We need Python for postprocessing the documentation
+  find_package(PythonInterp)
+  if (PYTHONINTERP_FOUND)
+    include(FindPythonModule)
+    find_python_module(bs4 FALSE)
+    find_python_module(pybtex FALSE)
+    if (PY_BS4 AND PY_PYBTEX)
+      # Construct the command that runs the postprocessing over the Doxygen HTML
+      # output
+      set(
+        DOCS_POST_PROCESS_COMMAND
+        "${PYTHON_EXECUTABLE} \
+${CMAKE_SOURCE_DIR}/docs/config/postprocess_docs.py \
+--html-dir ${PROJECT_BINARY_DIR}/docs/html \
+--references-file ${CMAKE_SOURCE_DIR}/docs/References.bib"
+      )
+      # Append postprocessing to doxygen commands
+      # The commands are supposed to run the postprocessing even if the doc
+      # generation failed with warnings, so that we output useful documentation
+      # in any case. The commands exit successfully only if both generation and
+      # postprocessing succeeded.
+      set(
+        GENERATE_DOCS_COMMAND
+        "${GENERATE_DOCS_COMMAND} && ${DOCS_POST_PROCESS_COMMAND} -v"
```

Comment:
isn't this `&&` the thing we were trying to avoid so that the next command would run even if the first would fail? Maybe I'm not remembering the previous code correctly. @wthrowe could you look/test? The goal of the changes @nilsleiffischer made was to generate the full shell script in memory and write it to disk once at the end, just need to be sure nothing changed in terms of how it works :)

I'm happy with the rest, just want to make sure this works as requested :)

Reply:
This does what you say @nilsdeppe. The `GENERATE_DOCS_COMMAND` is only running Doxygen, not checking for warnings, so it fails only if something is seriously wrong. It's for the `make doc` target. The `make doc-check` target does the file-writing and combines the commands in the way we discussed.

So both targets run Doxygen, then the postprocessing.